File: rustypad/src/leetcode/_41_word_in_all_sent.py
"""
Count of words that are present in all the given sentences

Given n sentences. The task is to count the number of words that appear in all of 
these sentences. 
Note that every word consists of only lowercase English alphabets.
"""
from typing import List
from collections import defaultdict


def get_top_words(documents: List[str], k: int):
    # - create a dictionary of word as key and documents index as value
    # - iterate each word and check if documents >= k then add to result
    bow = defaultdict(list)
    for doc_id, doc in enumerate(documents):
        for token in doc.split():
            bow[token].append(doc_id)
    results = [word for word, doc_ids in bow.items() if len(doc_ids) >= k]
    return results


def __get_top_words(docs: List[str], k: int) -> List[str]:
    """
    :param _type_ docs: _description_
    :param _type_ k: _description_
    :return _type_: _description_
    """
    from itertools import chain
    from collections import Counter
    token = chain(*[doc.split() for doc in docs])
    most_common = Counter(token).most_common(k)
    return [token for token, _ in most_common]


def main():
    arr = ["there is a cow",
           "cow is our mother",
           "cow gives us milk and milk is sweet",
           "there is a boy who loves cow"]
    K = len(arr)
    print(__get_top_words(arr, K))


if __name__ == "__main__":
    main()

File: rustypad/src/leetcode/_41_word_in_all_sent.rs
/* 
Count of words that are present in all the given sentences

Given n sentences. The task is to count the number of words that appear in all of 
these sentences. 
Note that every word consists of only lowercase English alphabets.

 */

use std::collections::HashMap;

#[derive(Debug)]
struct Solution;

impl Solution {

    /// - create a dictionary of word as key and documents index as value
    /// - iterate each word and check if documents >= k then add to result
    fn get_top_words(documents: Vec<&str>, k: i32) -> Vec<&str> {
        let mut results: Vec<&str> = Vec::new();
        let mut bow: HashMap<&str, Vec<usize>> = HashMap::new();
        for (doc_id, doc) in documents.iter().enumerate() {
            let tokens: Vec<&str> = doc.split(' ').collect();
            for token in tokens {
                match bow.get_mut(token) {
                    Some(bow_docs) => bow_docs.push(doc_id),
                    None => { bow.insert(token, vec![doc_id]); },
                }
            }
        }

        for (word, document_ids) in bow {
            if document_ids.len() >= k as usize {
                results.push(word)
            }
        }

        return results;
    }
}

#[test]
fn test() {
    let arr = vec!["there is a cow",
           "cow is our mother",
           "cow gives us milk and milk is sweet",
           "there is a boy who loves cow"];
    let k = arr.len();
    let ans = Solution::get_top_words(arr, k as i32);
}